### ISSUE TITLE
Jasmine 2.1 beforeAll/afterAll methods cause errors

### ISSUE THREAD

**Issue by chesleybrown** [angular/angular.js#10238]
Jasmine has released `beforeAll` and `afterAll` methods with [2.1](http://jasmine.github.io/2.1/introduction.html) which allow executing a `before` function only once for all the following specs. Angular mocks doesn't seem to play nice with this concept.

When using `beforeAll` or `afterAll` methods, I encounter the following errors.

```
TypeError: 'null' is not an object (evaluating 'currentSpec.$modules')
	    at workFn (/Users/chesleybrown/Sites/bln-web/app/components/angular-mocks/angular-mocks.js:2323)
	    at /Users/chesleybrown/Sites/bln-web/node_modules/karma-jasmine/lib/boot.js:71
	    at /Users/chesleybrown/Sites/bln-web/node_modules/karma-jasmine/lib/adapter.js:174
	    at http://localhost:9876/karma.js:185
	    at http://localhost:9876/context.html:168
	TypeError: 'null' is not an object (evaluating 'currentSpec.$modules')
	    at workFn (/Users/chesleybrown/Sites/bln-web/app/components/angular-mocks/angular-mocks.js:2323)
	    at /Users/chesleybrown/Sites/bln-web/node_modules/karma-jasmine/lib/boot.js:71
	    at /Users/chesleybrown/Sites/bln-web/node_modules/karma-jasmine/lib/adapter.js:174
	    at http://localhost:9876/karma.js:185
	    at http://localhost:9876/context.html:168
	Expected { 0: HTMLNode, length: 1, prevObject: { 0: HTMLNode, length: 1 }, context: undefined, selector: '.loading-indicator' } to have class 'ng-hide'.
	Error: Expected { 0: HTMLNode, length: 1, prevObject: { 0: HTMLNode, length: 1 }, context: undefined, selector: '.loading-indicator' } to have class 'ng-hide'.
	    at /Users/chesleybrown/Sites/bln-web/test/directives/about-product.directive.spec.js:40
	    at /Users/chesleybrown/Sites/bln-web/node_modules/karma-jasmine/lib/boot.js:71
	    at /Users/chesleybrown/Sites/bln-web/node_modules/karma-jasmine/lib/adapter.js:174
	    at http://localhost:9876/karma.js:185
	    at http://localhost:9876/context.html:168
```

**Comment by lgalfaso** [comment]
After two months and no feedback, I made the assumption that the issue is what @shahata commented. Closing this, but if there is anything else that needs to be handled, then feel free to reopen the issue

**Comment by scriby** [comment]
The use case I have for using beforeAll looks something like this:

```
describe('some expensive to create object', function() {
  var object;

  beforeAll(function() {
    object = createExpensiveObject();
  });

  it('has property A', function() {
    expect(object).toHavePropertyA();
  });

  it('has property B', function() {
    expect(object).toHavePropertyB();
  });

  ...
});
```

I like being able to split up the assertions into more granular chunks instead of just listing them one after another in a single it block.

I think this is still achievable with beforeEach (just initialize object if it's null), it's just not quite so straightforward.

**Comment by xmlilley** [comment]
So... is the bottom line that Angular doesn't support Jasmine's `beforeAll`, and developers should avoid it? In my case, I spent an hour chasing unclear errors, only to eventually track down this issue...

**Comment by lgalfaso** [comment]
It is not possible to use the injector at `beforeAll` as there is no cross-application nor cross multiple bootstrap injector. If there is some need to make Jasmine work in a different way, I would recommend you to open an issue at the karma-jasmine plugin repo

**Comment by timruffles** [comment]
@lgalfaso not sure why this would be an Jasmine issue? Jasmine's `beforeAll` (and mocha's `before`) works fine, it's the implementation of `angular-mocks` that stops it working for angular tests.

As @xml said, seems a shame to talk about Jasmine as the most popular framework in the docs, and not support a very core feature for writing good tests in the style @scriby demonstrated.

**Comment by lgalfaso** [comment]
@timruffles each test must have its own clean `$injector`. This is the only way to warranty that the right configuration is present for each test. This implies that anything that needs an injector must be executed once per test.
Jasmine `beforeAll` syntax is very powerful, but given that it runs once for all tests, then it cannot have access to the `$injector` (and subsequently it cannot be used in conjunction with `inject`).

Loose comments stating this is a *shame* do not help. The project *karma-jasmine* adds some functionality to *Jasmine* to make it possible for it to be used within *karma* to test *Angular* apps. My comment above does not claim that there is a need to change how *Jasmine* works, it is a great library that works in the expected way. That does not mean that all the functionality provided by *Jasmine* can be used with the functionality provided by *karma-jasmine*. One of the limitation is that you do not have access to `$injector` at `beforeAll`. If you find a solution to this problem, then open a PR on *karma-jasmine* and someone will review it.

**Comment by timruffles** [comment]
@lgalfaso sometimes code is a better way to explain.

[Here is a repo](https://github.com/timruffles/angular-jasmine-with-before-all) which demonstrates how it is possible to support this functional changing only ngMock. This demonstrates it's ngMock's jasmine/mocha support that is the limiting factor here. 

Rather than tying injector state to `currentSpec`, a `currentContext` is [created on demand](https://github.com/timruffles/angular-jasmine-with-before-all/blob/master/ng-mock-jasmine-integration-that-supports-before-all.js#L81), and [cleared on demand](https://github.com/timruffles/angular-jasmine-with-before-all/blob/master/ng-mock-jasmine-integration-that-supports-before-all.js#L19).

As you can see, it supports [beforeEach/beforeAll and the standard use of injectors](https://github.com/timruffles/angular-jasmine-with-before-all/blob/master/test/mainTest.js).

If you would like this feature in ngMock I'd be happy to write the code, update the tests etc. I'm a GDE so I'd be glad to get started with contribution, rather than just doing talks etc.

**Comment by lgalfaso** [comment]
@timruffles can you make your variation of ngMock work with this simple currently working example

```js
describe('something', function() {
  beforeEach(module(function($interpolateProvider) {
    $interpolateProvider.startSymbol('[[');
    $interpolateProvider.endSymbol(']]');
  }));

  it('should interpolate using [[ ]]', inject(function($interpolate) {
    expect($interpolate('[[foo]]')({foo: 'bar'})).toBe('bar');
  }));
});

describe('something else', function() {
  it('should interpolate using {{ }}', inject(function($interpolate) {
    expect($interpolate('{{foo}}')({foo: 'bar'})).toBe('bar');
  }));
});
```

**Comment by timruffles** [comment]
Thanks for the example, yes I can! I added [opt in](https://github.com/timruffles/angular-jasmine-with-before-all/commit/cd8b003f33330de929ddbffda3b69f7f9526b7b3) so you can selectively turn on the new behaviour for given tests. Your example passes without modification, I updated my test with the opt in `  beforeAll(module.enableManualInjector)`.

Run `karma start modifiedNgMock.conf.js` to see your tests passing successfully, alongside tests using beforeAll.

**Comment by lgalfaso** [comment]
Good, I think we are going somewhere, but at the end the main point is not clear.
1. Existing test, just like the one in the example, must work unmodified
2. Adding `beforeAll(inject(...))` should work with the existing examples

This is, the following code should work
```js
beforeAll(inject(function($http) {}));

// Paste here the previous example code
```

**Comment by timruffles** [comment]
Great.

1. This is achieved - your example works unmodified.
1. `beforeAll(inject(function($http) {}))` doesn't work with the current version of angular/angular-mock. [1]

I'm not sure what the intention is in supporting `inject()` without `module()`? The use cases people are asking about above are all using `beforeAll()` with `module()` then `inject()`.

[1] e.g, 

```javascript
beforeAll(inject(function($http) {}));

describe('something', function() {
  beforeEach(module(function($interpolateProvider) {
    $interpolateProvider.startSymbol('[[');
    $interpolateProvider.endSymbol(']]');
  }));

  it('should interpolate using [[ ]]', inject(function($interpolate) {
    expect($interpolate('[[foo]]')({foo: 'bar'})).toBe('bar');
  }));
});

describe('something else', function() {
  it('should interpolate using {{ }}', inject(function($interpolate) {
    expect($interpolate('{{foo}}')({foo: 'bar'})).toBe('bar');
  }));
});
```

throws this with the current `master` version of angular-mock:

```
Chrome 46.0.2490 (Mac OS X 10.9.5) something should interpolate using [[ ]] FAILED
	TypeError: Cannot read property '$modules' of null
	    at Object.workFn (/Users/timruffles/dev/oss/angular-jasmine-with-before-all/node_modules/angular-mocks/angular-mocks.js:2413:32)
Chrome 46.0.2490 (Mac OS X 10.9.5) something else should interpolate using {{ }} FAILED
	TypeError: Cannot read property '$modules' of null
	    at Object.workFn (/Users/timruffles/dev/oss/angular-jasmine-with-before-all/node_modules/angular-mocks/angular-mocks.js:2413:32)
```

**Comment by timruffles** [comment]
In case you're reading in email, I edited the comment above with an example of the use cases for `beforeAll()`.

**Comment by timruffles** [comment]
@lgalfaso I'm up for writing this up into a PR if we're agreed on beforeAll support via opt-in.

**Comment by lgalfaso** [comment]
I was not able look at your approach, should have some time tomorrow. That
said, I am not fond on having an opt-in behavior

**Comment by lgalfaso** [comment]
@timruffles I am reading your comment on the proposal and I am not sure how you are proposing this to work. Let me write down what my understanding on your proposal is, and if this is not correct, please write the comprehensive idea.

When there is a `beforeAll` that makes use of `$injector` then all the tests in the `describe` context that `beforeAll` is, will share a unique `$injector`. If there is a global `beforeAll` that makes use of `$injector`, then all the tests share this unique `$injector`.

Eg.
```js
describe('X', function() {
  // Here, each test has its own $injector.
  it('X1', inject(function(...) {}));
  it('X2', inject(function(...) {}));
  it('X3', inject(function(...) {}));
});

describe('Y', function() {
  // Here, all test share the same $injector.
  beforeAll(inject(function(...) {}));

  it('Y1', inject(function(...) {}));
  it('Y2', inject(function(...) {}));
  it('Y3', inject(function(...) {}));
});
```

**Comment by timruffles** [comment]
Really it's not tied to `beforeAll` at all! Instead we allow users to control when an injector is created, which will enable a lot of good testing practises currently not supported.

The implementation is exposing a `enableManualInjector` switch. This happens to works very well with `beforeAll()`, and supports the typical use cases for `beforeAll()` that people were discussing above:

- clean tests: one action, many expectations
- speeding up slow tests, where one expensive action is taken and many assertions are made about it

A complete use-case would be:

```javascript
describe('some expensive to create object', function() {
  var object;

  beforeAll(module.enableManualInjector)
  beforeAll(module("someModule"))

  afterAll(module.disableManualInjector)

  beforeAll(inject(function(createExpensiveObject) {
    object = createExpensiveObject();
  }));

  it('has property A', function() {
    expect(object).toHavePropertyA();
  });

  it('has property B', function() {
    expect(object).toHavePropertyB();
  });

});
```

**Comment by lgalfaso** [comment]
@timruffles I think this conversation is not moving forward. Exposing a test configuration that has global side-effects and that depends on the *good will* of developers to undo this configuration, is a bad practice. Test should be isolated and independent **by default** and must be impossible for some configuration in another file, left behind by some sloppy developer or just because some other test did (not) run before to affect the stability or result of any other test that did not explicitly opted-in.
This is about the minimum that anybody expects from a testing library.

Now, there is a need to allow `beforeAll` to perform some expensive operation and for the result of this expensive operation to be shared between multiple tests. The result of allowing this would improve the speed that tests are executed. This is, allowing a group of tests to be coupled and interdependent with each other to gain some speed. Anyhow, this behavior must not affect any other test, and this must be regardless if the developer added `afterAll(module.disableManualInjector())` or not.

Any proposal must follow these constraints. If you are willing to work on a PR within these constraints, then I am ok in continuing with this discussion. If you still think that following these constraints are not what you have in mind, then I would feel more comfortable with some other core member team to continue on my behalf.

**Comment by timruffles** [comment]
Great points. I've updated the solution to expose only one method: `module.sharedInjector()`. This shares an injector for the current describe block only - leaving the rest of the suite unaffected.

That avoids the danger you mentioned where other tests would be affected, and doesn't require developers to manually cleanup.

Please let me know if there are other constraints I've missed. Otherwise I'll take a look at that PR.

```javascript
describe('allowing shared injectors, without affecting other tests', function() {

  describe('tests with sharedInjector', function() {
    module.sharedInjector();

    // within here, we've got 1 injector across all beforeEach, tests etc
    beforeAll(module("main", function($provide) {
      $provide.value("stubMe", {
        service: "stubbed",
      })
    }))
  })

  describe('tests with isolation', function() {
    // within here, we've got isolated injectors again

    beforeEach(module("main", function($provide) {
      $provide.value("stubMe", {
        service: "stubbed",
      })
    }))

  })
});
```

**Comment by timruffles** [comment]
@lgalfaso hopefully I've addressed your points above? Was just reminded of this work by [this related issue](https://github.com/karma-runner/karma-mocha/issues/72) on karma-mocha being closed.

**Comment by lgalfaso** [comment]
@timruffles sorry for taking this long to answer, I know I still have to look into this. At the API level, I think this is fine, but did not have the time to look at the code/test/docs. Hopefully will have some time next week

**Comment by petebacondarwin** [comment]
Hello @timruffles and @lgalfaso - I am late to this discussion but have just popped in.
It seems to me that it would be better if we could just get `beforeAll(inject(...))` working by itself without the developer having to "know" about the `manualInjector` switch. This would follow the path of least surprise and be a better experience for developers in general.
Can we not identify if `inject` was called in `beforeAll` and if so then create a shared context for all the `it` clauses but also automatically define an `afterAll` call to remove this shared injector at the end of the tests?

**Comment by gkalpak** [comment]
It would still be a big surprize if all of a sudden my `it` blocks start having a shared injector.
So, if all this is happening behind the scenes, it should be very clearly documented imo.

**Comment by timruffles** [comment]
Good point @gkalpak, but as `before()/beforeAll()` (mocha/jasmine) are both currently broken when used with `module()`, I don't think there can be any existing users to be surprised.

**Comment by petebacondarwin** [comment]
@gkalpak it would only be all of a sudden if you actually had a `beforeAll(inject(...))` block already defined.
How about banning the use of `inject(...)` if a shared injector has been defined already? A clear error message (e.g. "You cannot use `inject` here as a shared injector has already been defined in a previous `beforeAll` block). That would make it very clear to the programmer and would minimize surprise.
The developer using `beforeAll` would then be responsible for moving services into closure variables to ensure they were available to other `beforeEach` and `it` blocks, which I think would be  reasonable compromise.

**Comment by gkalpak** [comment]
_"all of a sudden"_ was not in the sense of a BC, but more like:

> 1. I've been writing test in Angular for several months. Everything works fantastic. I know that a new injector is created for every `it` block, so I don't have to clear up after each.
> 2. Aha, new shiny `beforeAll` feature. Let's put that into the mix.
> 3. Ooops, tests break, because of dirty state left over form previous tests.

Banning the use of `inject(...)` won't solce this issue I think.

--
All I'm saying is that going from "1 injector per block" to "1 injector for all blocks" is not explicitly clear just because I used a different method and it should be clearly documented (be it one the dev guide, the changelog or wherever appropriate).

**Comment by petebacondarwin** [comment]
Do you have any suggestions to improve the situation?

**Comment by gkalpak** [comment]
All I am saying is, if we will have it work automagically (i.e. without the user's having to opt-in with something like the `manualInjector` switch) - which is fine by me btw - we should clearly document the implications of using `beforeAll()`, e.g. in the [Unit Testing section of the guide][1] and also mention them in the commit message, so it is also mentioned in the changelog.
(I.e. not just say: _Yay, added support for `beforeAll()`_, but something like _Yay, added support for `beforeAll()`. If you use it, this and that will happen._)

That's all :smiley: 

[1]: https://docs.angularjs.org/guide/unit-testing

**Comment by petebacondarwin** [comment]
How about an explicit API:  `beforeAll(sharedInject(...));`?

**Comment by gkalpak** [comment]
It sounds pretty good to me. (But you still have to document it - there's no escaping that :stuck_out_tongue:)

**Comment by timruffles** [comment]
`beforeAll(module.sharedInjector("someModule", function($provide) {}))` certainly works for me as an API!

There are decisions to be made about interacting with non-shared injector code. I think for safety it's best to disallow this:

```
describe
  beforeEach1 module
  T1
  describe 
     beforeAll1 module.sharedInjector
     T2
```

This could technically work, as we'd see the following calls:

```
beforeEach1, T1
beforeAll1, beforeEach1, T2
```

However, I imagine a lot of people will get into nasty situations where their `beforeEach`'s in outer contexts assume they always have an isolated injector. The ordering of the `beforeAll` means that wouldn't be the case, and the same beforeEach code would run in isolated and shared contexts.

So for safety I think we should only support `beforeAll(module.sharedInjector` in a clean context. There's no way to opt out of it within a context, so you can't nest isolated within shared. e.g

```
// ok
beforeEach module.sharedInjector
describe
  beforeAll inject()

// not ok
beforeEach inject()
describe
  beforeEach module.sharedInjector
```

As desiring a shared injector is probably the exception rather than the rule, this restriction seems ok!

If this looks ok with you @petebacondarwin, I'll rebase my work into a PR.

**Comment by bluecollarcoder** [comment]
Just ran into this problem. Was this issue resolved with v1.6.1?

**Comment by timruffles** [comment]
[See the docs here](https://docs.angularjs.org/guide/unit-testing#using-beforeall-) on how to use `beforeAll()`.

**Comment by bluecollarcoder** [comment]
@timruffles Thanks!

**Comment by gkalpak** [comment]
(This has been fixed since v1.5.1 btw.)
